### ISSUE TITLE
Fallback not working as expected

### ISSUE THREAD

**Issue by wwortel** [aler9/rtsp-simple-server#225]
v 0.14.2
an audio stream 'purpose' is being published at intervals. When 'purpose' is not being published those connecting to 'purpose' should hear an announcement 'back_soon'. 
Back_soon is being published using ffmepg from Init onwards and defined under paths: as fallbackpath: and under path 'purpose:' as 'fallback: ....' entry. 
Observed behaviour:
when 'purpose' is not being published and one connects to 'purpose' for listening, back_soon is streamed : OK
when then a publisher starts to publish 'purpose', it has no effect on the listener. The listener keeps hearing back_soon: FALSE
when one connects to 'purpose' while 'purpose' is being published one hears 'purpose': OK
when then 'purpose is interrupted the clients stops hearing anything and the client's stream is dropped: FALSE

The desired behaviour is that the 'purpose' is always of higher priority than the fallback. So, when fallback is taking the place of 'purpose' in absence of 'purpose', the stream given out should be swapped back for 'purpose' once 'purpose' becomes available without interrupting the stream. The client should be unaware of any of this.

**Comment by aler9** [comment]
Hello, at the moment the fallback system works by redirecting clients to the fallback URL, exactly like a HTTP redirect. A redirect doesn't allow to go back.

Anyway, let's suppose to implement the behavior you're proposing: let's suppose we send to the clients, in an invisible way, the frames of a fallback stream, and switch back to the original one as soon as it is published. The problem is that in RTSP the stream description (i.e. how many tracks and what are their codecs) is sent once, at the beginning of the stream.
If the fallback stream and the original streams have two different codecs... it can't be done, and the stream must be re-initialized.

If someone has a better idea on how to implement this, i'd like to hear it

**Comment by golfterminator** [comment]
I am intrigued by your suggestion of silently replacing the frames of the publishing stream until it resumes publishing.  This exactly the behavior that I am looking for.

I understand it is tricky to handle the general case where the fallback and original stream have different codec/attributes.

However, if, in my application , I can ensure that the fallback stream is the same codec/attributes, would that be a straight forward change to just send the fallback stream frames instead?

**Comment by wwortel** [comment]
@aler9, thanks for comments and for all your work.
The 'back_soon' message is coded exactly as the 'purpose'. It is just a recording in endless loop and one can conveniently choose its format. In my case both the 'purpose' and 'back_soon' are fixed rate (32k) mp3. But 'back_soon' can always be adapted to whatever format 'purpose' is published in. So indeed the question is whether one can seamlessly swap one for the other, or even fill a time gap with silence until the decision is made to switch back to 'back_soon' when 'purpose' has been absent for to long, without the client noticing any of it. The track number and attributes would not have to be adapted as both contents elements are about 'purpose', either 'purpose' itself or its announcement.

**Comment by golfterminator** [comment]
I've been playing around with the fallback feature and I've added more logging to help me understand how things work.  

System starts up and both my real starem "4HE3-1" and the fallback stream "idlepath" are publishing.

When I play the "4HE3-1" from VLC it works fine and "4HE3-1" is playing.

When I stop the "4HE3-1" stream, I see the "4HE3-1" path being cleaned up and removed:

> 2021/03/07 11:51:48 [I] [3/2/1] [clientrtsp 127.0.0.1:59860] [readDone]
> 2021/03/07 11:51:48 [I] [3/2/1] [path 4HE3-1] <-pa.clientRemove
> 2021/03/07 11:51:48 [I] [3/2/1] [path 4HE3-1] removeClient 127.0.0.1 state 3
> 2021/03/07 11:51:48 [I] [3/1/1] [path 4HE3-1] removeClient 172.16.3.88 state 1
> 2021/03/07 11:51:48 [I] [3/1/0] [path 4HE3-1] <-pa.closeTimer.C
> 2021/03/07 11:51:48 [I] [3/1/0] [PM] <-pm.pathClose pa.Name 4HE3-1
> 2021/03/07 11:51:48 [I] [3/1/0] [path 4HE3-1] Done
> 2021/03/07 11:51:48 [I] [3/1/0] [path 4HE3-1] cleaning up client 172.16.3.88
> 2021/03/07 11:51:48 [I] [2/1/0] [clientrtsp 172.16.3.88:51401] [terminate]
> 2021/03/07 11:51:48 [I] [2/1/0] [clientrtsp 172.16.3.88:51401] disconnected
> 2021/03/07 11:51:48 [I] [1/1/0] [clientrtsp 127.0.0.1:59860] disconnected

The VLC client is also terminated.  I see that VLC stops playing the stream.  After a few seconds, VLC starts playing the "idlepath" stream automatically:

> 2021/03/07 11:51:56 [I] [2/1/0] [clientrtsp 172.16.3.88:51403] connected (plain)
> 2021/03/07 11:51:56 [I] [2/1/0] [clientrtsp 172.16.3.88:51403] [c->s] rtsp://172.16.2.115:554/idlepath
> 2021/03/07 11:51:56 [I] [2/1/0] [clientrtsp 172.16.3.88:51403] [s->c] 
> 2021/03/07 11:51:56 [I] [2/1/0] [clientrtsp 172.16.3.88:51403] [c->s] rtsp://172.16.2.115:554/idlepath
> 2021/03/07 11:51:56 [I] [2/1/0] [clientrtsp 172.16.3.88:51403] [onDescribe] idlepath
> 2021/03/07 11:51:56 [I] [2/1/0] [PM] findPathConf name idlepath
> 2021/03/07 11:51:56 [I] [2/1/0] [PM] <-pm.clientDescribe req.name idlepath pathName idlepath
> 2021/03/07 11:51:56 [I] [2/1/0] [path idlepath] <-pa.clientDescribe
> 2021/03/07 11:51:56 [I] [2/1/0] [path idlepath] onClientDescribe state 2

So far this is all proper behavior and the fallback stream is working.  But my first question os, how does VLC know to reconnect to "idlepath"?  Where is the code that tells the VLC client to redirect to "idlepath"?

I've also looked at the situation where the "4HE3-1" path is not publishing yet and only the "idlepath" is publishing.  When I connect to "4HE3-1", I see that in the onClientDescribe() func in path/path.go it redirects the client when the state is sourceStateNotReady:

>     case sourceStateNotReady:
>         if pa.conf.Fallback != "" {
>             fallbackURL := func() string {
>                 if strings.HasPrefix(pa.conf.Fallback, "/") {
>                     ur := base.URL{
>                         Scheme: req.Req.URL.Scheme,
>                         User:   req.Req.URL.User,
>                         Host:   req.Req.URL.Host,
>                         Path:   pa.conf.Fallback,
>                     }
>                     return ur.String()
>                 }
>                 return pa.conf.Fallback
>             }()

This code is not hit in the case the "4HE3-1" path is stopped and VLC is redirected to "idlepath".  So there must be another code path for that redirect?

My second question is, can the same mechanism be used to force VLC to reconnect to "4HE3-1" when it resumes publishing?

Thanks
Dave

**Comment by riccardocagnasso** [comment]
> v 0.14.2
> an audio stream 'purpose' is being published at intervals. When 'purpose' is not being published those connecting to 'purpose' should hear an announcement 'back_soon'.
> Back_soon is being published using ffmepg from Init onwards and defined under paths: as fallbackpath: and under path 'purpose:' as 'fallback: ....' entry.
> Observed behaviour:
> when 'purpose' is not being published and one connects to 'purpose' for listening, back_soon is streamed : OK
> when then a publisher starts to publish 'purpose', it has no effect on the listener. The listener keeps hearing back_soon: FALSE
> when one connects to 'purpose' while 'purpose' is being published one hears 'purpose': OK
> when then 'purpose is interrupted the clients stops hearing anything and the client's stream is dropped: FALSE
> 
> The desired behaviour is that the 'purpose' is always of higher priority than the fallback. So, when fallback is taking the place of 'purpose' in absence of 'purpose', the stream given out should be swapped back for 'purpose' once 'purpose' becomes available without interrupting the stream. The client should be unaware of any of this.

You can accomplish what you want with some gstreamer magic.

https://pastebin.com/4KWeXt3a

**Comment by wwortel** [comment]
Hi Riccardo, thanks for your suggestion and code submitted. I assume it is totally obvious to you, but I would need a little more explanation to be able to try your code. How and where does this fit into rtsp-simple-server? Your code is in Python (that I am more familiar with than with Go) but I'd need some more clues to figure out how this 'magic' could be integrated to get the seamless switching between intermittent real-time primary published stream and some filler recorded stream in absence of the primary. Or is this a stand-alone solution?

**Comment by riccardocagnasso** [comment]
> 
> 
> Hi Riccardo, thanks for your suggestion and code submitted. I assume it is totally obvious to you, but I would need a little more explanation to be able to try your code. How and where does this fit into rtsp-simple-server? Your code is in Python (that I am more familiar with than with Go) but I'd need some more clues to figure out how this 'magic' could be integrated to get the seamless switching between intermittent real-time primary published stream and some filler recorded stream in absence of the primary. Or is this a stand-alone solution?

Basically I take one stream from the server, unpack the h264 from the rtsp and put it back in the server in another stream.
Then it's gstreamer magic with probes and event to detect if the stream is stopped, delink the old rtsp and link a new one for the same or another stream. You can manipulate this to do what you like, like changing the stream when you want or when something is disconnected etc.

But... seamless... well not really. I'm queuing some h264 after some other h264. There will be a choppy transition between them. If you want true seamless, you have to decode and reencode and use videomixer.

**Comment by aler9** [comment]
> there will be a choppy transition between them

Exactly, the standard way to do this consists in using a server with a broadcasting software installed (like OBS Studio for instance) and handle transitions from there, it's not much different from what TV channels do
